lib/zhuyin.ex

```diff
@@ -0,0 +1,326 @@
+   "r"
+  """
+  # @spec decode_zhuyin(String.t()) :: Integer.t()
+   def decode_zhuyin(input) do
+    case input |> Pinyin.ZhuYinParsers.zhuyin_word |> IO.inspect do
+      {:ok, parsed, _rest, _other1, _other2, _other3} ->
+        first = parsed |> List.first
+        if first == input do
+          None
+        else
+          first |> Zhuyin.to_pinyin |> Pinyin.numbered
+        end
+      true -> None
+    end
+   end
+
+  @spec to_pinyin(t()) :: Pinyin.t()
+  # Special case for this final and tone combination
+  def to_pinyin(%Zhuyin{initial: "", final: "ㄦ", tone: 0}) do
+    Pinyin.create("", "r", 0)
+  end
+  def to_pinyin(zhuyin = %Zhuyin{}) do
+    pinyin_initial = if zhuyin.initial == "" do
+      ""
+    else
+      @initials[zhuyin.initial]
+    end
+    pinyin_final = if zhuyin.initial == "" do
+      @standalone_finals[zhuyin.final]
+    else
+      @finals[zhuyin.final]
+    end
+    Pinyin.create(pinyin_initial, pinyin_final, zhuyin.tone)
+  end
+  def to_pinyin(list) when is_list(list) do
+    list |> Enum.map(&Zhuyin.to_pinyin/1)
+  end
+
+  @spec to_zhuyin(Pinyin.t() | Pinyin.pinyin_list()) :: t()
+  def to_zhuyin(pinyin = %Pinyin{}) do
+    initial_map = Map.new(@initials, fn {key, val} -> {val, key} end)
+    initial = initial_map[pinyin.initial]
+
+    pinyin.final |> IO.inspect
+    final = if pinyin.initial == "" do
+      standalone_finals_map = Map.new(@standalone_finals, fn {key, val} -> {val, key} end)
+      standalone_finals_map[pinyin.final]
+    else
+      finals_map = Map.new(@finals, fn {key, val} -> {val, key} end)
+      finals_map[pinyin.final]
+    end
+
+    %__MODULE__{initial: initial, final: final, tone: pinyin.tone}
+  end
+  def to_zhuyin(list) when is_list(list) do
+    list
+    |> Enum.map(fn
+      p = %Pinyin{} -> to_zhuyin(p)
+      str when is_binary(str) -> str
+    end)
+    |> Enum.join()
+  end
+
+  def create(initial, final, tone) do
+    %__MODULE__{initial: initial, final: final, tone: tone_index(tone)}
+  end
+  def create(final, tone) do
+    %__MODULE__{final: final, tone: tone_index(tone)}
+  end
+
+  @spec read(String.t(), :exclusive | :words | :mixed) ::
+          {:ok, zhuyin_list()} | {:error, String.t()}
+  def read(string, mode \\ :exclusive) when mode in [:exclusive, :words, :mixed] do
+    res =
+      case mode do
+        :exclusive -> Pinyin.ZhuYinParsers.zhuyin_only(string)
+        :words -> Pinyin.ZhuYinParsers.zhuyin_words(string)
+        :mixed -> Pinyin.ZhuYinParsers.mixed_words(string)
+      end
+
+    case res do
+      {:ok, lst, "", %{}, _, _} -> {:ok, lst}
+      {:error, _, remainder, %{}, _, _} -> {:error, remainder}
+    end
+  end
+
+  @spec read!(String.t(), :exclusive | :words | :mixed) ::
+          zhuyin_list() | no_return()
+  def read!(string, mode \\ :exclusive)
+      when mode in [:exclusive, :words, :mixed] do
+    case read(string, mode) do
+      {:ok, res} -> res
+      {:error, remainder} -> raise Pinyin.ParseError, remainder
```

Comment:
Perhaps you can create a `Zhuyin.ParseError` to avoid confusion.

Reply:
I assume this naming would be shown in the error message? Okay, done.

Reply:
I've renamed `Pinyin.ParseError` to just `ParseError`, so you can use that, so that we don't have to maintain two separate  yet similar exception modules.

Reply:
Ok, I'll use that.